### PR TITLE
Proposed Roadmap for torch.sparse matrix multiplication 

### PR DESCRIPTION
**Pull request by dhavide**
Suggestions for a strategy on how to go forward with `torch.sparse`

- This doc discusses only matrix multiplication (because it is mentioned in many issues)
- The goal is to start here as a way to go deeper into other issues associated with `torch.sparse`.

### CODE REVIEW

**Comment by ezyang** [comment]
One very basic newbie question: is there a generalization of CSR format that works for higher dimensions? Or is it a 2D only kind of deal?

**Comment by gchanan** [comment]
I did a quick read, thanks for the thorough writeup!

Some questions that come to mind:
1) Can the csr format be generalized to more than just 2-d tensors?  Having "2d" in all the names feels pretty clunky and we could just error out if the input matrix is not 2d if it's not generalizable.
2) I'd like to see more detail on the constructor(s). The conversion operator IMO doesn't really matter because it's mostly used in tests (most uses for sparse tensors you wouldn't want to materialize the dense tensor anyway).

**Comment by hameerabbasi** [comment]
> One very basic newbie question: is there a generalization of CSR format that works for higher dimensions? Or is it a 2D only kind of deal?

There is GCRS (https://ieeexplore.ieee.org/document/7237032) and CSF (https://dl.acm.org/doi/10.1145/2833179.2833183). If you want to go completely-bonkers-general-all-formats-are-one there's TACO (https://arxiv.org/abs/1804.10112).

**Comment by albanD** [comment]
Hi,

Thanks for the proposal.
I would have one question about the timeline.
I assume that this new Tensor type will have `.is_sparse() == true`. But how is it going to interact with code that already exists and use this to special case on the COO sparse Tensors we have?
For example here https://github.com/pytorch/pytorch/blob/5bbcddae3bfb7abaeaada29369c6ed5964d390fe/torch/csrc/autograd/functions/accumulate_grad.h#L61-L68 we assume that when we have a sparse Tensor, then we can access _values and _indices. But this might not be true anymore.
Should a new point be added to the timeline to investigate the current codebase and make sure all the special casing we do for sparse Tensors will be ok for all sparse Tensors?

**Comment by dhavide** [comment]
> I did a quick read, thanks for the thorough writeup!
> 
> Some questions that come to mind:
> 
>     1. Can the csr format be generalized to more than just 2-d tensors?  Having "2d" in all the names feels pretty clunky and we could just error out if the input matrix is not 2d if it's not generalizable.
> 
>     2. I'd like to see more detail on the constructor(s). The conversion operator IMO doesn't really matter because it's mostly used in tests (most uses for sparse tensors you wouldn't want to materialize the dense tensor anyway).

1. As Hameer points out, there are generalizations beyond 2D. We actually do need to consider this
    because there are use cases (e.g., NLP) in which storing batches of sparse matrices is useful.
    The precise implementation of this is quite complicated and needs to be considered carefully.
    We can modify the proposal to make this more explicit.

    Anyway, it's fine with to change the identifiers (someone else has already winced at the `2d`
    suffix for the modifications to the API). The primary motivation was to serve as a warning of a
    restriction, but that restriction may not exist in the eventual implementation.
2. More details about the constructor is a good idea. We'll modify the proposal accordingly.

**Comment by dhavide** [comment]
> Hi,
> 
> Thanks for the proposal.
> I would have one question about the timeline.
> I assume that this new Tensor type will have `.is_sparse() == true`. But how is it going to interact with code that already exists and use this to special case on the COO sparse Tensors we have?
> For example here https://github.com/pytorch/pytorch/blob/5bbcddae3bfb7abaeaada29369c6ed5964d390fe/torch/csrc/autograd/functions/accumulate_grad.h#L61-L68 we assume that when we have a sparse Tensor, then we can access _values and _indices. But this might not be true anymore.
> Should a new point be added to the timeline to investigate the current codebase and make sure all the special casing we do for sparse Tensors will be ok for all sparse Tensors?

Excellent point. Yes, this has the potential to break existing parts of the codebase, so an additional point needs to be added to the timeline to verify this (concerns about `autograd` are mentioned in passing in the current draft, but this certainly needs to be called out explicitly in the timeline).

File: PyTorch-Sparse-Roadmap.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr2d` and `torch.to_sparse_coo` to convert sparse PyTorch
+   tensors of sparse rank 2 between COO and CSR formats.
+
+   + The implementation of related functions in 
+     [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul)
+     can serve as a useful model for these transformations.
+   + Include exception handling to ensure `sparse_dim()==2` before converting to from COO to CSR.
+   + See also comments in issue
+     [\#16187](https://github.com/pytorch/pytorch/issues/16187).
```
Comment (ezyang) [inline PyTorch-Sparse-Roadmap.md:272]: NB: conversion functions need to work on CUDA :>

File: PyTorch-Sparse-Roadmap.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+   + See [issue # 16187](https://github.com/pytorch/pytorch/issues/16187)
+     for representative examples of benchmarks.
+   + Include examples of matrices of fairly large dimensions (i.e., where dense
+     representations are impractical or infeasible).
+   + Include examples with `sparse x dense` and `sparse x sparse` matrices for
+     the timings.
+
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr2d` and `torch.to_sparse_coo` to convert sparse PyTorch
+   tensors of sparse rank 2 between COO and CSR formats.
```
Comment (ezyang) [inline PyTorch-Sparse-Roadmap.md:265]: If CSR is going to be another layout type, there is some infrastructural work that is going to need to be done here; you need a new tensor type (since a CSR isn't going to have the same memory representation as COO tensor). For example, right now we refer to COO sparse as just "Sparse" in the codebase; it might make sense to disambiguate these names to "SparseCOO" so people don't mix them up. (Or, I suppose, we can just grandfather sparse to mean COO sparse, and be explicit in all the new locations).

Comment (pearu) [reply PyTorch-Sparse-Roadmap.md:265]: The identity "Sparce == Sparce COO" is not consistent in current pytorch, for instance, there are sparse tools that include `coo` explicitly in their names: `torch.sparse_coo_tensor`, `at::_sparse_coo_tensor_with_dims`, `torch.sparse_coo`, etc.
This indicates that the design of the current infrastructure implicitly would support other sparse formats. For instance, one could have  `torch.sparse_csr_tensor`, etc.

**Comment by rgommers** [comment]
> > One very basic newbie question: is there a generalization of CSR format that works for higher dimensions? Or is it a 2D only kind of deal?
> 
> There is GCRS (https://ieeexplore.ieee.org/document/7237032) and CSF (https://dl.acm.org/doi/10.1145/2833179.2833183). If you want to go completely-bonkers-general-all-formats-are-one there's TACO (https://arxiv.org/abs/1804.10112).

Important disclaimer there is that those >2-D generalizations are not supported by cuSPARSE, MKL et al. CSR is the best-supported format by those libraries, followed by COO. So it's unclear to me whether it makes sense to have more than one >2-D format.

File: PyTorch-Sparse-Roadmap.md (hunk 1 above)
Comment (dhavide) [reply PyTorch-Sparse-Roadmap.md:272]: Noted. Appropriate modification made in revised proposal.

File: PyTorch-Sparse-Roadmap.md (hunk 2 above)
Comment (dhavide) [reply PyTorch-Sparse-Roadmap.md:265]: I think Pearu's comment is correct; it is possible to extend the current sparse classes without having to specify the storage format by a name like "SparseCOO." This could perhaps be achieved by adding a `_layout` attribute of type `torch.sparse_coo` or `torch.sparse_csr` or some other valid PyTorch object of class `torch.layout`.

The proposal has been modified to provide a more thorough description of the CSR format and the associated constructor.

File: PyTorch-Sparse-Roadmap.md [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+
+   + See [issue # 16187](https://github.com/pytorch/pytorch/issues/16187)
+     for representative examples of benchmarks.
+   + Include examples of matrices of fairly large dimensions (i.e., where dense
+     representations are impractical or infeasible).
+   + Include examples with `sparse x dense` and `sparse x sparse` matrices for
+     the timings.
+
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr2d` and `torch.to_sparse_coo` to convert sparse PyTorch
```
Comment (pearu) [inline PyTorch-Sparse-Roadmap.md:323]: Change: `to_sparse_csr2` -> `to_sparse_csr`

File: PyTorch-Sparse-Roadmap.md [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+  conversion between representations (for rank 2 sparse tensors).
+
+The `torch.sparse_csr_tensor` constructor would have a similar signature to `torch.sparse_coo_tensor`, i.e.,
+
+```{python}
+sparse_csr_tensor(indices, values, size=None, dtype=None,
+                  device=None, requires_grad=False)       -> Tensor
+```
+
+where `indices` would be a list of two lists (e.g., `row_index` & `col_index`
```
Comment (pearu) [inline PyTorch-Sparse-Roadmap.md:306]: Does the construction of CSR from `indices, values` allow N-D generalization? E.g. when using GCRS or CSF formats?
If yes, then +1. If not, what would be the uniting model for constructing CSR and GCRS?

File: PyTorch-Sparse-Roadmap.md [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+from the example before) and `values` would be a tensor containing the
+corresponding nonzero entries of the sparse tensor.
```
Comment (pearu) [inline PyTorch-Sparse-Roadmap.md:308]: Remove "nonzero" or replace it with "presumably nonzero" as nothing would check if the entries will be nonzero (it just would be too expensive to do that).

Should we be explicit here on whether the dense tensor of `values` can be an N-dimensional tensor o(similar to the COO case) or not?

File: PyTorch-Sparse-Roadmap.md (hunk 4 above)
Comment (pearu) [reply PyTorch-Sparse-Roadmap.md:306]: Or would it make sense to introduce a new layout for each generalization instead?

**Comment by ezyang** [comment]
I think we've already gotten here in the convo, but just to state it explicitly: if csr unambiguously refers to 2d only, we don't need to qualify the name with 2d; and we don't have to work on generalized versions if it's not clear, 2d only is fine.

**Comment by albanD** [comment]
Wouldn't a batch of sparse matrix be an important usecase as well? 
For nn use, have only 2D Tensors will be quite limited.

**Comment by dzhulgakov** [comment]
This is great proposal! Interesting thought: CSR is effectively equivalent to `offsets` representation that embedding_bag and other operations use. So it'd be good to explore whether we can change APIs there too to accept sparse tensors.

I'd also vote for extending representation to >2 dimensions if possible (or at least not baking it in API names now). Batch use case Alban mentioned is one example. Another point is that several embedding_bags' input can be thought as sparse 3d tensor (e.g. batch X num_features X feature_dim for sparse features). So it might be interesting to explore whether sparse tensor api is a good representation for it.

File: PyTorch-Sparse-Roadmap.md (hunk 3 above)
Comment (dhavide) [reply PyTorch-Sparse-Roadmap.md:323]: Whoops, fixed now. Thanks.

File: PyTorch-Sparse-Roadmap.md (hunk 4 above)
Comment (dhavide) [reply PyTorch-Sparse-Roadmap.md:306]: Okay, I've modified the description of CSR to N dimensions to include a specification of how to do so. Basically, for a sparse tensor of rank `d>=2` whose leading dimension has length `M`, the CSR format requires storing three arrays: `RO` (an integer array of length `M+1`), `CO` (an integer matrix of dimension `(d-1) x nnz`), and `VL` (an array of values of length `nnz`).

I think that this generalization is sufficient (even if the indexing sparse dimension is always the leading dimension). Permitting a new layout for other specific generalizations is available for those who choose to do so.

Representing batches of sparse matrices is required in numerous contexts (e.g., a large number of embedding matrices in NLP). The obvious Python way to do this would be in lists or dicts or standard containers, but, depending on the size of the batches, performance issues are an obvious problem.

File: PyTorch-Sparse-Roadmap.md (hunk 5 above)
Comment (dhavide) [reply PyTorch-Sparse-Roadmap.md:308]: Fixed.

**Comment by dhavide** [comment]
> Wouldn't a batch of sparse matrix be an important usecase as well?
> For nn use, have only 2D Tensors will be quite limited.

@albanD  yes, batches of sparse matrices is obviously important. I have modified the proposal to describe how CSR extends to tensors of arbitrary rank (and added references to papers that describe other extensions in greater detail). I believe this category of generalization is sufficient (even if the indexing sparse dimension in `RO` is always the leading dimension).

**Comment by pearu** [comment]
@dhavide could you also rename `PyTorch-Sparse-Roadmap.md` to `PyTorch-Sparse-MatMul-Roadmap.md`, or similar?

**Comment by jamestwebber** [comment]
@rgommers pointed me here from the CuPy issue–glad to see progress being made! To add some comments from an end-user (domain is computational biology).

> The big assumption is that faster sparse matrix products actually constitutes a meaningful benefit to PyTorch users—the hope is that the accuracy of this assumption will become clear in the early stages of prototyping.

I think this would have a major benefit in my area of research. I work a lot with single-cell RNA sequencing data (scRNAseq), and the raw data is a large, sparse count matrix. By "large": on the order of 10k rows for one experiment, easily 100k to 1M+ for very large projects or meta-analyses, and 10s of thousands of features. By "sparse": typically 1-2% nonzero values. Not "big data" so much as wide, complex data. This type of data is becoming more prevalent and datasets are growing as methods improve, see [this google sheet](https://docs.google.com/spreadsheets/d/1En7-UV0k0laDiIfjFkdn7dggyR7jIk3WH8QgXaMOZF0/edit#gid=1444925081) showing the trend over time (credit to @vals who might drop in later).

When analyzing scRNA data, the difference between sparse and dense representations can be the difference between fitting the data in GPU memory or not, which can greatly impact the speed of training. The rest of the model could be dense but just using sparse tensors for the input would allow for larger batches and less memory-copying. It's hard to know exactly how big a benefit that would provide but I think it'd be considerable.


One thing I would add as something to think about (but out of scope here, I guess): I'd love to be able to pass a sparse tensor in to `torch.distributions`. For example, call `torch.distributions.Binomial(total_counts=sparse_counts, p=0.5).sample()` and get a sparse tensor with the same shape and sampled counts in the nonzero locations.

That's kind of a niche thing to do but it might be doable with the `fill_value` attribute, for distributions where certain values have a trivial answer. My hunch is that most users want sparse tensors to be a drop-in replacement as much as possible, because trying to keep track of the options gets difficult (I always have this issue with scipy, where the sparse datatype is always a matrix, not an array).

File: RFC-0004-pyTorch-sparse-matmul-roadmap.md [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+The `torch.sparse_csr_tensor` constructor would have a similar signature to `torch.sparse_coo_tensor`, i.e.,
+
+```{python}
+sparse_csr_tensor(indices, values, size=None, dtype=None,
+                  device=None, requires_grad=False)       -> Tensor
+```
+
+where `indices` would be a list of two lists (e.g., `RO` & `CO`
+from the description of CSR format above) and `values` would be a tensor
+containing the corresponding (presumably) nonzero entries of the sparse
```
Comment (pearu) [inline RFC-0004-pyTorch-sparse-matmul-roadmap.md:332]: This API for CSR is incorrect/inefficient: `RO` has a length `size[0]+1` while `CO` has length `nnz` and so `RO` and `CO` lists should be separated.

I suggest the following API for `sparse_csr_tensor`:
```python
sparse_csr_tensor(pointers, indices, values, size=None, dtype=None,
                  device=None, requires_grad=False)       -> Tensor
```
where the `pointers` argument is a 1-D array of length size[0] + 1 containing the cumulative number of nonzero elements in the sparse array computed for each index of the first sparse dimension.

Xref: https://github.com/Quansight/pearu-issues/issues/28

File: RFC-0004-pyTorch-sparse-matmul-roadmap.md [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+   + See [issue # 16187](https://github.com/pytorch/pytorch/issues/16187)
+     for representative examples of benchmarks.
+   + Include examples of matrices of fairly large dimensions (i.e., where dense
+     representations are impractical or infeasible).
+   + Include examples with `sparse x dense` and `sparse x sparse` matrices for
+     the timings.
+
+2. Construct and test a prototype implementation of conversion functions
+   `torch.to_sparse_csr` and `torch.to_sparse_coo` to convert sparse PyTorch
+   tensors of sparse rank 2 between COO and CSR formats.
```
Comment (pearu) [inline RFC-0004-pyTorch-sparse-matmul-roadmap.md:351]: I think we can target immediately arbitrary rank (>=2) CSR sparse tensors in the constructor, including the hybrid feature, without any extra effort. The algorithms, however, may be restricted to 2-D in the short run.

File: RFC-0004-pyTorch-sparse-matmul-roadmap.md [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+respectively contain the extents of rows, the column indices, and the numerical
+values. That is, to represent an `M x N` sparse matrix `A` with `nnz` nonzero
+entries (floating-point) in CSR format requires three separate rank 1 arrays:
+`RO` (an integer array of length `M+1`), `CO` (an integer array of length
+`nnz`), and `VL` (a floating-point array of length `nnz`). In the
+extension of CSR format to tensors of rank greater `d>2`, the array `CO`
+would be a matrix of dimensions `(d-1) x nnz`.
+
+The entry in `RO[r]` is the cumulative number of nonzero entries in the matrix
+up to and including row `r` (`r=0...M+1`). Thus, the first entry of `RO` is `0`
```
[the remark this answers is not in the document]
Comment (pearu) [reply RFC-0004-pyTorch-sparse-matmul-roadmap.md]: Fixed.

File: RFC-0004-pyTorch-sparse-matmul-roadmap.md [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+and the last&mdash;`RO[M+1]`&mdash;is `nnz`. The arrays `CO` and `VL` of length
```
[the remark this answers is not in the document]
Comment (pearu) [reply RFC-0004-pyTorch-sparse-matmul-roadmap.md]: Fixed.

**Comment by npuichigo** [comment]
So will block sparse be supported as some people ask for that maybe for sparse attention? (https://github.com/pytorch/pytorch/issues/9222)

**Comment by rgommers** [comment]
Thanks for asking @npuichigo. We've discussed it briefly recently and it's clearly of interest. The first target for increasing sparse matrix performance at the moment is the GCS format: https://github.com/pytorch/pytorch/pull/44190. Considering the block sparse use cases and interest in https://github.com/pytorch/pytorch/issues/9222, it does seem like that's a good next candidate. I'll add a note to the tracking issue for sparse tensor improvements (https://github.com/pytorch/pytorch/issues/44634) to reassess the priority of block-sparse.
